tests/OperationProcessor.spec.ts

```diff
@@ -40,53 +48,230 @@ async function addBatchOfOneOperation (opBuf: Buffer, cas: Cas, transactionTime:
+        i + 1,   // transactionTime
+        0        // operation index
+        );
+      const newVersion = operationProcessor.process(updateOp);
+      expect(newVersion).not.toBeUndefined();
+      expect(newVersion!).toBe(getHash(updateOp));
+      versions.push(newVersion!);
+    }
+
+    // Check first(), last(), prev(), next() return expected outputs
+    for (let i = 0; i < versions.length; ++i) {
+      expect(await operationProcessor.first(versions[i])).toBe(versions[0]);
+      expect(await operationProcessor.last(versions[i])).toBe(versions[versions.length - 1]);
+
+      if (i === 0) {
+        expect(await operationProcessor.previous(versions[i])).toBeUndefined();
+      } else {
+        expect(await operationProcessor.previous(versions[i])).toBe(versions[i - 1]);
+      }
+      if (i === versions.length - 1) {
+        expect(await operationProcessor.next(versions[i])).toBeUndefined();
+      } else {
+        expect(await operationProcessor.next(versions[i])).toBe(versions[i + 1]);
+      }
+    }
+  });
+
+  it('should correctly process updates in reverse order', async () => {
+    const ops = new Array(createOp!);
+    const opHashes = new Array(getHash(createOp!));
+
+    const numberOfUpdates = 10;
+    for (let i = 0; i < numberOfUpdates; ++i) {
+      const mostRecentVersion = opHashes[i];
+      const updateOp = await addBatchOfOneOperation(createUpdateOperationBuffer(
+        mostRecentVersion),
+        cas,
+        i + 1,   // transaction Number
+        i + 1,   // transactionTime
+        0        // operation index
+        );
+      const updateOpHash = getHash(updateOp);
+      ops.push(updateOp);
+      opHashes.push(updateOpHash);
+    }
+
+    for (let i = numberOfUpdates ; i > 0 ; --i) {
+      const newVersion = operationProcessor.process(ops[i]);
+      expect(newVersion).toBeDefined();
+      expect(newVersion!).toBe(opHashes[i]);
+    }
+
+    // Check first(), last(), prev(), next() return expected outputs. Since
+    // if the OperationProcessor did not process the operations correctly
+    // some version defined by these operations would be "invalid" and we would
+    // not get the correct output below.
+    for (let i = 0; i < opHashes.length; ++i) {
+      expect(await operationProcessor.first(opHashes[i])).toBe(opHashes[0]);
+      expect(await operationProcessor.last(opHashes[i])).toBe(opHashes[opHashes.length - 1]);
+
+      if (i === 0) {
+        expect(await operationProcessor.previous(opHashes[i])).toBeUndefined();
+      } else {
+        expect(await operationProcessor.previous(opHashes[i])).toBe(opHashes[i - 1]);
+      }
+      if (i === opHashes.length - 1) {
+        expect(await operationProcessor.next(opHashes[i])).toBeUndefined();
+      } else {
+        expect(await operationProcessor.next(opHashes[i])).toBe(opHashes[i + 1]);
+      }
+    }
+  });
+
+  it('should correctly process updates in every (5! = 120) order', async () => {
+    const ops = new Array(createOp!);
```

Comment:
Consider generalizing the setup portion and verification portion of the code so similar code doesn't repeat 3 times.